scripts/apps/desks/services/DesksFactory.js

```diff
@@ -180,7 +180,7 @@ export function DesksFactory($q, api, preferencesService, userList, notify,
                 return $q.when(self.userDesks);
             }
 
-            return this.fetchCurrentDeskId() // make sure there will be current desk
+            return $q.when().then(this.fetchCurrentDeskId()) // make sure there will be current desk
```

Comment:
Why is `$q.when().then` necessary? Doesn't `this.fetchCurrentDeskId()` already return a promise?

Reply:
Yep, it does return the promise. I used it to get the promise resolved immediately, as on pasting the url or refreshing the page `userDesks` were not available.

Reply:
I don't get it. If the result of the promise is not available yet, you can't do anything, but ensure that the calling function can wait for the promise.

If you resolve it immediately aren't you resolving it with an empty value?

Reply:
Not resolving with empty values. I just saw it [here](https://github.com/superdesk/superdesk-client-core/blob/2691fb450e25cf7eda1e38b2663ed3fc55868f6b/scripts/apps/desks/services/DesksFactory.js#L130) and used it and worked for me.

Reply:
I wanted to check the issue myself but couldn't reproduce it. Can you reproduce it consistently yourself? If so, how?

Reply:
Use sd-master as the server. Open any published item. Put a debugger here https://github.com/superdesk/superdesk-client-core/blob/master/scripts/apps/authoring/authoring/services/AuthoringService.js#L570 and refresh the page. You will see `self.userDesks = []`.

Reply:
You are not solving it at the right place Gyan.

Let's look at how the issue happens. As you said `self.userDesks` is empty at AuthoringService.js:570 (line numbers according to master `15ab2131`)

Let's investigate where `self.userDesks` is populated - there is only one additional usage in the file at line 45.

```
desks.fetchCurrentUserDesks().then((desksList) => {
    self.userDesks = desksList;
});
```

`self.userDesks` is populated asynchronously, but allows synchronous access which means race conditions will happen and is exactly the issue behind the ticket.

To solve this properly we must ensure that data which is fetched asynchronously can't be accessed synchronously. There are two ways to achieve that:
1. Make all functions who access `self.userDesks` asynchronous too, so they can wait for the promise to resolve.
2. Remove `self.userDesks` from the file entirely and ensure it is passed as an argument to the functions which required it. This way you could keep methods synchronous.

Reply:
@tomaskikutis fixed it.

Reply:
> To solve this properly we must ensure that data which is fetched asynchronously can't be accessed synchronously.

The underlying issue is still there. If you decided to go with option 2 and pass `userDesks` as an argument, you need to remove `self.userDesks` from the file entirely and pass `userDesks` to all other functions which depend on it.

Reply:
`authoring.itemActions` function is called at more than 50 places across 15 files. So, every time the `self.userDesks` is required, that is why we need to keep it there.

Reply:
Let's fix it in the 50 places then. That's why I provided 2 options: to make the methods themselves async or provide an argument. If you prefer I can do it.

Reply:
"If you prefer I can do it" 
I didn't get you?

Reply:
If you want I can take over this task from you and update the usages in 50+ places you mentioned.

Reply:
Ok if you wish you can take it over.